### PR TITLE
Interceptor + Resolver

### PR DESCRIPTION
**Pull request by sstanculeanu**
- Added interceptedValidatorInfo, interceptedValidatorInfoDataFactory, validatorInfoInterceptorProcessor, validatorInfoResolver

### CODE REVIEW

File: dataRetriever/resolvers/validatorInfoResolver.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	defer res.throttler.EndProcessing()
+
+	rd, err := res.parseReceivedMessage(message, fromConnectedPeer)
+	if err != nil {
+		return err
+	}
+
+	switch rd.Type {
+	case dataRetriever.HashType:
+		return res.resolveHashRequest(rd.Value, rd.Epoch, fromConnectedPeer)
```
Comment (iulianpascalau) [inline dataRetriever/resolvers/validatorInfoResolver.go:110]: not for this PR: we can implement request by hashes in order to optimize the requests/response mechanism. Will be made just as the transactions resolver

File: dataRetriever/resolvers/validatorInfoResolver_test.go [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+	}
+}
+
+func createMockValidatorInfo() state.ValidatorInfo {
+	return state.ValidatorInfo{
+		PublicKey: []byte("provided pk"),
+		ShardId:   123,
+		List:      string(common.EligibleList),
+		Index:     10,
+		Rating:    10,
```
Comment (iulianpascalau) [inline dataRetriever/resolvers/validatorInfoResolver_test.go:40]: maybe change this to 11 in order to have all fields hold different values?

File: process/interceptors/factory/interceptedValidatorInfoDataFactory_test.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		t.Parallel()
+
+		_, cryptoComponents := createMockComponentHolders()
+		args := createMockArgument(nil, cryptoComponents)
+
+		ividf, err := NewInterceptedValidatorInfoDataFactory(*args)
+		assert.Equal(t, process.ErrNilCoreComponentsHolder, err)
+		assert.True(t, check.IfNil(ividf))
+	})
+	t.Run("nil marshalizer should error", func(t *testing.T) {
```
Comment (iulianpascalau) [inline process/interceptors/factory/interceptedValidatorInfoDataFactory_test.go:43]: nil marshaller?

File: process/interceptors/factory/interceptedValidatorInfoDataFactory.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import (
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go-core/hashing"
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/process/peer"
+)
+
+type interceptedValidatorInfoDataFactory struct {
+	marshalizer      marshal.Marshalizer
```
Comment (iulianpascalau) [inline process/interceptors/factory/interceptedValidatorInfoDataFactory.go:12]: marshaller?

File: dataRetriever/resolvers/validatorInfoResolver.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	logger "github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/ElrondNetwork/elrond-go/storage"
+)
+
+// ArgValidatorInfoResolver is the argument structure used to create a new validator info resolver instance
+type ArgValidatorInfoResolver struct {
+	SenderResolver       dataRetriever.TopicResolverSender
+	Marshalizer          marshal.Marshalizer
```
Comment (iulianpascalau) [inline dataRetriever/resolvers/validatorInfoResolver.go:19]: marshaller?

File: process/interceptors/processor/validatorInfoInterceptorProcessor.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+)
+
+// ArgValidatorInfoInterceptorProcessor is the argument structure used to create a new validator info interceptor processor
+type ArgValidatorInfoInterceptorProcessor struct {
+	Marshalizer       marshal.Marshalizer
+	ValidatorInfoPool storage.Cacher
+}
+
+type validatorInfoInterceptorProcessor struct {
+	marshalizer       marshal.Marshalizer
```
Comment (iulianpascalau) [inline process/interceptors/processor/validatorInfoInterceptorProcessor.go:18]: marshaller?

File: process/interceptors/processor/validatorInfoInterceptorProcessor.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	if check.IfNil(args.ValidatorInfoPool) {
+		return process.ErrNilValidatorInfoPool
+	}
+
+	return nil
+}
+
+// Validate checks if the intercepted data can be processed
+func (viip *validatorInfoInterceptorProcessor) Validate(data process.InterceptedData, _ core.PeerID) error {
+	_, ok := data.(interceptedValidatorInfo)
```
Comment (iulianpascalau) [inline process/interceptors/processor/validatorInfoInterceptorProcessor.go:48]: we could easily drop this as it is done in the Save function anyway.

File: process/interceptors/processor/validatorInfoInterceptorProcessor_test.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+func createMockArgValidatorInfoInterceptorProcessor() processor.ArgValidatorInfoInterceptorProcessor {
+	return processor.ArgValidatorInfoInterceptorProcessor{
+		Marshalizer:       testMarshalizer,
+		ValidatorInfoPool: testscommon.NewCacherStub(),
+	}
+}
+
+func TestNewValidatorInfoInterceptorProcessor(t *testing.T) {
+	t.Parallel()
+
```
Comment (iulianpascalau) [inline process/interceptors/processor/validatorInfoInterceptorProcessor_test.go:51]: marshaller?

File: process/peer/interceptedValidatorInfo.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+		return nil, err
+	}
+
+	return validatorInfo, nil
+}
+
+// CheckValidity checks the validity of the received validator info
+func (ivi *interceptedValidatorInfo) CheckValidity() error {
+	// Verify string properties len
+	err := verifyPropertyLen(publicKeyProperty, ivi.validatorInfo.PublicKey)
```
Comment (iulianpascalau) [inline process/peer/interceptedValidatorInfo.go:78]: maybe inject a size for the required len
This should be 96

File: process/peer/interceptedValidatorInfo_test.go [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+		args := createMockArgInterceptedValidatorInfo()
+		args.NodesCoordinator = &shardingMocks.NodesCoordinatorStub{
+			GetValidatorWithPublicKeyCalled: func(publicKey []byte) (validator nodesCoordinator.Validator, shardId uint32, err error) {
+				return nil, 0, expectedErr
+			},
+		}
+
+		ivi, _ := NewInterceptedValidatorInfo(args)
+		require.False(t, check.IfNil(ivi))
+
```
Comment (SebastianMarian) [inline process/peer/interceptedValidatorInfo_test.go:123]: Remove empty line here and at L138

File: process/peer/interceptedValidatorInfo_test.go [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+	str := ivi.String()
+	assert.True(t, strings.Contains(str, fmt.Sprintf("pk=%s", logger.DisplayByteSlice(ivi.validatorInfo.PublicKey))))
+	assert.True(t, strings.Contains(str, fmt.Sprintf("shard=%d", validatorInfo.ShardId)))
+	assert.True(t, strings.Contains(str, fmt.Sprintf("list=%s", validatorInfo.List)))
+	assert.True(t, strings.Contains(str, fmt.Sprintf("index=%d", validatorInfo.Index)))
+	assert.True(t, strings.Contains(str, fmt.Sprintf("tempRating=%d", validatorInfo.TempRating)))
+	assert.True(t, strings.Contains(str, fmt.Sprintf("rating=%d", validatorInfo.Rating)))
+}
+
```
Comment (SebastianMarian) [inline process/peer/interceptedValidatorInfo_test.go:202]: Unit tests should be added?

File: process/peer/interceptedValidatorInfo.go [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+	if err != nil {
+		return err
+	}
+	err = verifyPropertyLen(listProperty, []byte(ivi.validatorInfo.List))
```
Comment (iulianpascalau) [inline process/peer/interceptedValidatorInfo.go:82]: this might be checked against a size of 128

File: process/peer/interceptedValidatorInfo.go [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+	if err != nil {
+		return err
+	}
+	err = verifyPropertyLen(rewardAddressProperty, ivi.validatorInfo.RewardAddress)
```
Comment (iulianpascalau) [inline process/peer/interceptedValidatorInfo.go:86]: this will be maximum 32 bytes

File: process/peer/interceptedValidatorInfo_test.go [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+		t.Parallel()
+
+		args := createMockArgInterceptedValidatorInfo()
+		args.DataBuff = nil
+
+		ivi, err := NewInterceptedValidatorInfo(args)
+		assert.Equal(t, process.ErrNilBuffer, err)
+		assert.True(t, check.IfNil(ivi))
+	})
+	t.Run("nil marshalizer should error", func(t *testing.T) {
```
Comment (iulianpascalau) [inline process/peer/interceptedValidatorInfo_test.go:44]: marshaller

File: process/peer/interceptedValidatorInfo_test.go [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+	str := ivi.String()
+	assert.True(t, strings.Contains(str, fmt.Sprintf("pk=%s", logger.DisplayByteSlice(ivi.validatorInfo.PublicKey))))
+	assert.True(t, strings.Contains(str, fmt.Sprintf("shard=%d", validatorInfo.ShardId)))
+	assert.True(t, strings.Contains(str, fmt.Sprintf("list=%s", validatorInfo.List)))
+	assert.True(t, strings.Contains(str, fmt.Sprintf("index=%d", validatorInfo.Index)))
+	assert.True(t, strings.Contains(str, fmt.Sprintf("tempRating=%d", validatorInfo.TempRating)))
+	assert.True(t, strings.Contains(str, fmt.Sprintf("rating=%d", validatorInfo.Rating)))
+}
+
+func TestInterceptedValidatorInfo_Identifiers(t *testing.T) {
```
Comment (iulianpascalau) [inline process/peer/interceptedValidatorInfo_test.go:203]: empty tests?

File: process/peer/interceptedValidatorInfo_test.go (hunk 11 above)
Comment (sstanculeanu) [reply process/peer/interceptedValidatorInfo_test.go]: forgot to delete the lines as all getters are tested in `TestInterceptedValidatorInfo_Getters`

File: dataRetriever/resolvers/validatorInfoResolver.go (hunk 1 above)
Comment (sstanculeanu) [reply dataRetriever/resolvers/validatorInfoResolver.go:111]: somehow I understood from the meeting that we'll only need by hash. Added a TODO for a further PR.

File: process/peer/interceptedValidatorInfo_test.go (hunk 15 above)
Comment (sstanculeanu) [reply process/peer/interceptedValidatorInfo_test.go]: forgot to delete the lines as all getters are tested in `TestInterceptedValidatorInfo_Getters`

File: process/peer/interceptedValidatorInfo_test.go (hunk 11 above)
Comment (SebastianMarian) [reply process/peer/interceptedValidatorInfo_test.go]: ok

**Review by SebastianMarian — APPROVED** [review]

File: dataRetriever/resolvers/validatorInfoResolver.go [hunk 16]
```diff
@@ -105,18 +105,14 @@ func (res *validatorInfoResolver) ProcessReceivedMessage(message p2p.MessageP2P,
 		return err
 	}
 
+	// TODO: add support for HashArrayType
 	switch rd.Type {
 	case dataRetriever.HashType:
 		return res.resolveHashRequest(rd.Value, rd.Epoch, fromConnectedPeer)
 	default:
```
Comment (iulianpascalau) [inline dataRetriever/resolvers/validatorInfoResolver.go:112]: useless default :D

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**